circuit_breaker.go

```diff
@@ -0,0 +1,159 @@
+package resty
+
+import (
+	"errors"
+	"net/http"
+	"sync/atomic"
+	"time"
+)
+
+// CircuitBreaker can be in one of three states: Closed, Open, or Half-Open.
+// When the circuit breaker is Closed, requests are allowed to pass through.
+// If a failure count threshold is reached within a specified time-frame, the circuit breaker transitions to the Open state.
+// When the circuit breaker is Open, requests are blocked.
+// After a specified timeout, the circuit breaker transitions to the Half-Open state.
+// When the circuit breaker is Half-Open, a single request is allowed to pass through.
+// If that request fails, the circuit breaker returns to the Open state.
+// If SuccessThreshold requests succeed, the circuit breaker transitions back to the Closed state.
+type CircuitBreaker struct {
+	policies                        []CircuitBreakerPolicy
+	timeout                         time.Duration
+	failThreshold, successThreshold int
+
+	state                   circuitBreakerState
+	failCount, successCount int
```

Comment:
These int should be atomic Int

https://pkg.go.dev/sync/atomic#Int64

The ++ while having concurrent http resource could break everything I think

Reply:
Very good catch! Done.